File: reference/src/rarity.ts
// 🔒 レア度計算（docs/02）。
import { u32 } from "./hash.ts";
import { RARITY_CUTOFFS_U32, TAGS, eraBonusFromYear } from "./constants.ts";

/**
 * 基本レア度 ★1..10。整数しきい値で判定（docs/08 8.4、f64非依存）。
 * 余り（どの cutoff にも満たない領域）は ★1（案A・実効40.49%）。
 */
export function baseRarity(seed: Uint8Array): number {
  const x = u32(seed, TAGS.rarity);
  for (let i = 0; i < RARITY_CUTOFFS_U32.length; i++) {
    if (x < RARITY_CUTOFFS_U32[i]!) return i + 1;
  }
  return 1; // 余り → ★1
}

/**
 * QR年 → 時代補正。
 * year=null（取得不可）/ 1900未満（不正）→ +0。範囲外(未来)は表側で +0。
 */
export function eraBonusFor(year: number | null): number {
  if (year === null) return 0;
  if (year < 1900) return 0; // 不正値は「年代取得不可」扱い（docs/07 Q8）
  return eraBonusFromYear(year);
}

export function clampRarity(n: number): number {
  return Math.max(1, Math.min(13, n));
}

/** 最終レア度 = clamp(base + bonus, 1, 13)。 */
export function finalRarity(base: number, bonus: number): number {
  return clampRarity(base + bonus);
}

/** ★11..13（神話級）。eraBonus≥1 のときのみ到達しうる。 */
export function isMythic(final: number): boolean {
  return final >= 11;
}

File: rust/qrac-core/src/rarity.rs
//! 🔒 レア度計算（docs/02）。整数しきい値（docs/08 8.4）。
use crate::constants::{era_bonus_from_year, RARITY_CUTOFFS_U32, TAG_RARITY};
use crate::hash::u32v;

/// 基本レア度 ★1..10。余りは ★1（案A）。
pub fn base_rarity(seed: &[u8]) -> u32 {
    let x = u32v(seed, TAG_RARITY, 0);
    for (i, &cut) in RARITY_CUTOFFS_U32.iter().enumerate() {
        if x < cut {
            return (i + 1) as u32;
        }
    }
    1
}

/// QR年 → 時代補正。null / 1900未満 → +0（docs/07 Q1/Q8）。
pub fn era_bonus_for(year: Option<i32>) -> u32 {
    match year {
        None => 0,
        Some(y) if y < 1900 => 0,
        Some(y) => era_bonus_from_year(y),
    }
}

pub fn final_rarity(base: u32, bonus: u32) -> u32 {
    (base + bonus).clamp(1, 13)
}

pub fn is_mythic(final_rarity: u32) -> bool {
    final_rarity >= 11
}
